### ISSUE TITLE
Explore Question Answering Model for Highlighting Patient Reviews

### ISSUE THREAD

**Issue by janduplessis883** [janduplessis883/ai-medreview#7]
![Image](https://github.com/user-attachments/assets/0a59933b-6901-4431-9ce3-ca99c1db7c67)

## Description:
To improve the analysis of patient reviews, we need to explore and evaluate a Question Answering (QA) model that can extract key insights from free-text feedback. The goal is to use NLP to identify and highlight relevant sections of patient comments based on specific queries, such as “What outcome does the patient want?” or “What feedback is given about waiting times?”.

## Objectives:
- Investigate pre-trained QA models (e.g., distilbert-base-cased-distilled-squad, bert-large-uncased-whole-word-masking-finetuned-squad).
- Test performance on real patient reviews, focusing on extracting meaningful insights.
- Evaluate accuracy and relevance of answers for different question types.
- Consider fine-tuning the model if pre-trained options don’t perform optimally.
- Assess efficiency and feasibility of integrating this approach into our feedback analysis pipeline.

## Tasks:
- Select a pre-trained transformers-based QA model (Hugging Face, etc.).
- Implement a basic QA pipeline using transformers library.
- Run initial tests on sample patient reviews.
- Evaluate extracted answers based on relevance and accuracy.
- Document findings and suggest next steps (fine-tuning, alternative models, etc.).

## Resources:
- [Hugging Face QA Models](https://huggingface.co/models?pipeline_tag=question-answering)
- Example patient reviews dataset (TBD)

## Priority: 
- 🔹 Medium

### Labels:
- NLP, QA Model, Patient Feedback, Research

### Next Steps:
- Conduct a feasibility assessment based on initial results.
- Determine if additional preprocessing, embeddings, or model fine-tuning are required for better performance.

**Comment by janduplessis883** [comment]
Transformer **Pipeline**

```python
from transformers import pipeline

qa_model = pipeline("question-answering", "timpal0l/mdeberta-v3-base-squad2")
question = "What outcome does the patient want?"
context = """Following a strange call from a member of your staff regarding my prescriptions which have been supplied to me through the Astell Pharmacy for at least 20 years I attempted to email Dr PERSON without success On having the email bounce back I then POSTED FIRST CLASS COPIES OT THE EMAIL to BOTH the Flood Street and Chelsea Manor Street addresses over a week ago As yet no response from Dr PERSON or someone on his behalf I have a MESSAGE service so there is no excuse"""
qa_model(question = question, context = context)
# {'score': 0.975547730922699, 'start': 28, 'end': 36, 'answer': ' Sweden.'}
```

**Update processed data**

```python
import numpy as np
from tqdm import tqdm
import pandas as pd

outcome_list = []

for _, row in tqdm(data.iterrows(), desc="QA", total=data.shape[0]):
    question = "What outcome does the patient want?"
    context = row.get('free_text', '')
    
    # Check if context is a valid non-empty string
    if not isinstance(context, str) or context.strip() == "":
        outcome_list.append(np.nan)
    else:
        try:
            result = qa_model(question=question, context=context)
            outcome_list.append(result)
        except Exception as e:
            print(f"Error processing row: {e}")
            outcome_list.append(np.nan)
```
